pkg/controller/applicationlayer/applicationlayer_controller.go

```diff
@@ -0,0 +1,319 @@
+	if err = r.client.Patch(ctx, applicationLayer, preDefaultPatchFrom); err != nil {
+		log.Error(err, "Failed to write defaults")
+		r.status.SetDegraded("Failed to write defaults", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	variant, installation, err := utils.GetInstallation(ctx, r.client)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.status.SetDegraded("Installation not found", err.Error())
+			return reconcile.Result{}, err
+		}
+		r.status.SetDegraded("Error querying installation", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	pullSecrets, err := utils.GetNetworkingPullSecrets(installation, r.client)
+
+	if err != nil {
+		r.status.SetDegraded("Error retrieving pull secrets", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	lcSpec := applicationLayer.Spec.LogCollection
+
+	// if ApplicationLayer spec exists then LogCollection should be set
+	// TODO: when we will have multiple features in future this should change to at least one feature being set
+	if lcSpec == nil {
+		r.status.SetDegraded(fmt.Sprintf("Error retrieving LogCollection Spec"), "")
+		return reconcile.Result{}, err
+	}
+
+	err = r.patchFelixTproxyMode(ctx, lcSpec)
+
+	if err != nil {
+		r.status.SetDegraded("Error patching felix configuration", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	if r.enableL7LogsCollection(lcSpec) {
+
+		l7component := applicationlayer.ApplicationLayer(pullSecrets, installation, rmeta.OSTypeLinux,
+			true, lcSpec.LogIntervalSeconds, lcSpec.LogRequestsPerInterval)
+
+		ch := utils.NewComponentHandler(log, r.client, r.scheme, applicationLayer)
+
+		if err = imageset.ApplyImageSet(ctx, r.client, variant, l7component); err != nil {
+			reqLogger.Error(err, "Error with images from ImageSet")
+			r.status.SetDegraded("Error with images from ImageSet", err.Error())
+			return reconcile.Result{}, err
+		}
+
+		// TODO: when there are more ApplicationLayer options then it will need to be restructured.
+		// because each of the different features will not have their own CreateOrUpdateOrDelete
+		if err := ch.CreateOrUpdateOrDelete(ctx, l7component, r.status); err != nil {
+			r.status.SetDegraded("Error creating / updating resource", err.Error())
+			return reconcile.Result{}, err
+		}
+
+	}
+
+	// Clear the degraded bit if we've reached this far.
+	r.status.ClearDegraded()
+
+	if !r.status.IsAvailable() {
+		// Schedule a kick to check again in the near future. Hopefully by then
+		// things will be available.
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	// Everything is available - update the CRD status.
+	applicationLayer.Status.State = operatorv1.TigeraStatusReady
+	if err = r.client.Status().Update(ctx, applicationLayer); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func validateApplicationLayer(al *operatorv1.ApplicationLayer) error {
+	lcSpec := al.Spec.LogCollection
+	if lcSpec != nil {
+		if *lcSpec.LogIntervalSeconds < 1 {
```

Comment:
@tmjd can't we get k8s to do this kind of validation by defining constraints in our CRDs, or am I mistaken?

Reply:
@Brian-McM We can, I just realised this, added it in this 971b9aa2427b8525d7bf163202d41274c57ff06f